docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+
+```swift
+public struct AsyncResponse {
+  /// the used http version
+  public var version: HTTPVersion
+  /// the http response status
+  public var status: HTTPResponseStatus
+  /// the response headers
+  public var headers: HTTPHeaders
+  /// the response payload as an AsyncSequence
+  public var body: Body
+}
+
+extension AsyncResponse {
+  public struct Body: AsyncSequence {
+    public typealias Element = ByteBuffer
+    public typealias AsyncIterator = Iterator
+  
+    public struct Iterator: AsyncIteratorProtocol {
+      public typealias Element = ByteBuffer
+    
+      public func next() async throws -> ByteBuffer?
+    }
+  
+    public func makeAsyncIterator() -> Iterator
+  }
+}
+```
+
+At a later point we could add trailers to the AsyncResponse as effectful properties.
+
+```swift
+    public var trailers: HTTPHeaders { async throws }
+```
+
+However we will need to make sure that the user has consumed the body stream completely before, calling the trailers, because otherwise we might run into a situation from which we can not progress forward:
+
+```swift
+do {
+  var request = AsyncRequest(url: "https://swift.org/")
+  let response = try await httpClient.execute(request, deadline: .now() + .seconds(3))
+  
+  var trailers = try await response.trailers // can not move forward since body must be consumed before.
+} catch {
+  print(error)
+}
+
+```
+
+### New invocation
+
+The new way to invoke a request shall look like this:
+
+```swift
+extension HTTPClient {
+  func execute(_ request: AsyncRequest, deadline: NIODeadline) async throws -> AsyncResponse
+}
+```
+
+- **Why do we have a deadline in the function signature?** 
+    Task deadlines are not part of the Swift 5.5 release. However we think that they are an important tool to not overload the http client accidentally. For this reason we will not default them.
+- **What happened to the Logger?** We will use Task locals to propagate the logger metadata. @slashmo and @ktoso are currently working on this.
+- **How does cancellation work?** Cancellation works by cancelling the surrounding task:
+
+        ```swift
+        let task = Task {
+          let response = try await httpClient.execute(request, deadline: .distantFuture)
+        }
+        
+        Task.sleep(500 * 1000 * 1000) // wait half a second
+        task.cancel() // cancel the task after half a second
+        ```
+
+- **What happens with all the other configuration options?** Currently users can configure a TLSConfiguration on a request.  This API doesn't expose this option. We hope to create a three layer model in the future. For this reason, we currently don't want to add per request configuration on the request invocation. More info can be found in the issue: [RFC: design suggestion: Make this a "3-tier library"][issue-392]
```

Comment:
Hah yeah sorry I promised to get to that one as welll... will try to make the time. But agreed on the more layers approach in general!